model/DataStore/PersistDataService.php

```diff
@@ -0,0 +1,142 @@
+    private const ITEM_META_DATA_JSON = 'itemMetaData.json';
+    private const PACKAGE_FILENAME = 'QTIPackage';
+    private const ZIP_EXTENSION = '.zip';
+
+    public function persist(array $params): void
+    {
+        $fileSystem = $this->getDataStoreFilesystem();
+        $folder = $this->getFolderName($params['deliveryId']);
+
+        $this->persistData($fileSystem, $folder, self::DELIVERY_META_DATA_JSON, $params['deliveryMetaData']);
+        $this->persistData($fileSystem, $folder, self::TEST_META_DATA_JSON, $params['testMetaData']);
+        $this->persistData($fileSystem, $folder, self::ITEM_META_DATA_JSON, $params['itemMetaData']);
+
+        $this->persistExportedTest($params['deliveryId'], $params['testUri']);
+    }
+
+    /**
+     * @throws common_exception_Error
+     * @throws common_exception_NotFound
+     */
+    private function getDataStoreFilesystem(): FileSystem
+    {
+        return $this->getFileSystemManager()->getFileSystem(self::DATA_STORE);
+    }
+
+
+    private function getFolderName(string $deliveryId): string
+    {
+        return tao_helpers_Uri::encode($deliveryId);
+    }
+
+    private function persistData(FileSystem $fileSystem, string $folder, string $fileName, $params): void
+    {
+        if (!$fileSystem->has($folder . DIRECTORY_SEPARATOR . $fileName)) {
+            $fileSystem->write($folder . DIRECTORY_SEPARATOR . $fileName, json_encode($params));
+        }
+    }
+
+    private function persistExportedTest(string $deliveryId, string $testUri): void
+    {
+        /** @var FileHelperService $tempDir */
+        $tempDir = $this->getServiceLocator()->get(FileHelperService::class);
+        $folder = $tempDir->createTempDir();
+
+        $testExporter = $this->getTestExporter();
+        try {
+            $testExporter->export(
+                [
+                    'filename' => self::PACKAGE_FILENAME,
+                    'instances' => $testUri,
+                    'uri' => $testUri
+                ],
+                $folder
+            );
+
+            $this->moveExportedZipTest($folder, $deliveryId, $tempDir);
+        } catch (Throwable $exception) {
+            $this->logError(
+                'DataStore: An error has occurred while exporting the qti package ::' .
+                $exception->getMessage()
+            );
+        }
+    }
+
+    /**
+     * @throws common_exception_Error
+     * @throws common_exception_NotFound
+     */
+    private function moveExportedZipTest(string $folder, string $deliveryId, FileHelperService $tempDir): void
+    {
+        $zipFiles = glob(
+            sprintf('%s%s*%s', $folder, self::PACKAGE_FILENAME, self::ZIP_EXTENSION)
+        );
+
+        if (!empty($zipFiles)) {
+            foreach ($zipFiles as $zipFile) {
+                $this->logDebug('Started to copy zip file: ' . $zipFile);
+                $contents = file_get_contents($zipFile);
+                $this->getDataStoreFilesystem()->write(
```

Comment:
In case of retry, will it override the file or throw an error? Cause IMO it should not block the process if the file is already there. Wdyt?